### DISCUSSION TITLE
How to implement retrying strategy for rate-limits?

### DISCUSSION THREAD

**Discussion by cmpadden** [MarshalX/atproto#468] [Questions] [answered]
Hello!

Is there a recommended approach for handling rate limiting using out-of-the-box features? I was hoping to use `httpx.Retry`, however,  I don't see an obvious way of propagating an `httpx.Retry` via `kwargs` to the `ClientBase._invoke` method through the client or the parameters of `get_author_feed`.

```
retry_strategy = httpx.Retry(
    total=3,
    status_forcelist=[500, 502, 503, 504],
    backoff_factor=0.5,
    method_whitelist=["GET"],
)

client.get(url, retry=retry_strategy)
```

```
return self.app.bsky.feed.get_author_feed(
    models.AppBskyFeedGetAuthorFeed.Params(actor=actor, cursor=cursor, filter=filter, limit=limit)
)
```

Thank you.

**Comment by MarshalX** [comment]
Hello! `**kwargs` should be passed directly to the request`s one. But not from the high level client, yes. 

So use more low level one:
```python
client = Client()....
client.app.bsky.feed.get_author_feed(
    models.AppBskyFeedGetAuthorFeed.Params(actor=actor, cursor=cursor, filter=filter, limit=limit),
    retry=httpx.Retry(.......),
)
```

Unfortunately, I do not see available `retry` arg in `httpx_client.request()`. But you can try

**Comment by cmpadden** [comment]
Thanks @MarshalX .

You're right that the `httpx.Retry` I shared above seems to be from an older version, and not supported.

I'll explore what other options there are, and update in this thread if/when I find something. Thank you.

**Comment by cmpadden** [comment]
For others, I ended up using `tenacity`, as I couldn't figure out how to inject a transport into the client, which the `httpx` documentation suggests for retries.
```
>>> transport = httpx.HTTPTransport(retries=1)
>>> client = httpx.Client(transport=transport)
```

Seems to work well enough!

```
@tenacity.retry(
    stop=tenacity.stop_after_attempt(5),
    wait=tenacity.wait_fixed(math.ceil(60 * 2.5)),
)
def _get_feed_with_retries(client: Client, actor: str, cursor: Optional[str]):
    return client.get_author_feed(actor=actor, cursor=cursor, limit=100)

feed = []
cursor = None
while True:
    data = _get_feed_with_retries(client, actor, cursor)
    feed.extend(data.feed)
    cursor = data.cursor
    if not cursor:
        break
```

**Reply by MarshalX** [reply, marked as answer]
hi, checks this one. it will help you to pass your transport: [#467 (comment)](https://github.com/MarshalX/atproto/discussions/467?converting=1#discussioncomment-11384769)

**Reply by cmpadden** [reply]
Very nice, thank you @MarshalX.

**Comment by MarshalX** [comment]
Configuration has been simplified. More info: https://atproto.blue/en/latest/atproto_client/timeouts.html. But for your case I agree that you should use tenacity
